x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -1360,6 +1376,35 @@ private void initialize(ClusterService clusterService) {
         });
     }
 
+    private final AtomicBoolean installTokenMetadataCheck = new AtomicBoolean(false);
+
+    private void installTokenMetadata(MetaData metaData) {
+        if (metaData.custom(TokenMetaData.TYPE) == null) {
+            if (installTokenMetadataCheck.compareAndSet(false, true)) {
+                clusterService.submitStateUpdateTask("install-token-metadata", new ClusterStateUpdateTask(Priority.URGENT) {
+                    @Override
+                    public ClusterState execute(ClusterState currentState) throws Exception {
+                        XPackPlugin.checkReadyForXPackCustomMetadata(currentState);
+
+                        if (currentState.custom(TokenMetaData.TYPE) == null) {
+                            return ClusterState.builder(currentState).putCustom(TokenMetaData.TYPE, getTokenMetaData()).build();
+                        } else {
+                            return currentState;
+                        }
+                    }
+
+                    @Override
+                    public void onFailure(String source, Exception e) {
+                        installTokenMetadataCheck.set(false);
+                        logger.error("unable to install token metadata", e);
+                    }
+                });
+            }
+        } else {
+            installTokenMetadataCheck.set(false);
```

Comment:
I don't follow why this gets set to `false` if the custom metadata already exists.
It seems like it could introduces an extraordinarily unlikely race condition (probably impossible in practice):
1. `Thread 1`: custom metadata is null,
2. `Thread 1`: set installTokenMetadataCheck to `true`
3. `Thread 2`: custom metadata is null
4. `Thread 1`: install metadata
5. `Thread 3`: custom metadata is non-null
6. `Thread 3`: set installTokenMetadataCheck to false
7. `Thread 2`: set installTokenMetadataCheck to true 
8. `Thread 2`: try into install duplicate metadata.

Reply:
I think the pattern has been copied from other sections of code that want to be defensive against the possibility of their cluster state changes being deleted somehow, and wanting to reinstall them.  It's not really a problem if the unlikely race condition you describe occurs, as it doesn't hurt correctness to get into the `execute()` method twice, only performance.

The history is that there are similar bits of code in ML and Watcher that didn't have the atomic guard variable when first written.  So originally they might update the cluster state 100 times with the same change as the cluster started up.  After observing this happening we added the atomic guard and it stopped enough of the duplicate metadata additions that it's no longer a noticeable problem.

Reply:
Also this happens inside of a cluster state listener and IIRC these are all executed on the cluster state update thread, which prevents multi-threading issues around this.

Reply:
As @droberts195 and @jaymode explained, `installTokenMetadata` is called by a single thread, but possibly repeatedly so while the cluster state update task has not taken effect yet. Initially I copied the code from a similar task in ML, but have made a smaller simplification in 87ad80f214b that should make it clearer why and when the flag is reset.